tests/test_ratelimiter.py

```diff
@@ -0,0 +1,356 @@
+from telegram.request import BaseRequest, RequestData
+from tests.conftest import env_var_2_bool
+
+TEST_NO_RATE_LIMITER = env_var_2_bool(os.getenv("TEST_NO_RATE_LIMITER", False))
+
+
+@pytest.mark.skipif(
+    not TEST_NO_RATE_LIMITER, reason="Only relevant if the optional dependency is not installed"
+)
+class TestNoRateLimiter:
+    def test_init(self):
+        with pytest.raises(RuntimeError, match=r"python-telegram-bot\[rate-limiter\]"):
+            AIORateLimiter()
+
+
+class TestBaseRateLimiter:
+    rl_received = None
+    request_received = None
+
+    async def test_no_rate_limiter(self, bot):
+        with pytest.raises(ValueError, match="if a `ExtBot.rate_limiter` is set"):
+            await bot.send_message(chat_id=42, text="test", rate_limit_args="something")
+
+    async def test_argument_passing(self, bot_info, monkeypatch, bot):
+        class TestRateLimiter(BaseRateLimiter):
+            async def initialize(self) -> None:
+                pass
+
+            async def shutdown(self) -> None:
+                pass
+
+            async def process_request(
+                self,
+                callback,
+                args,
+                kwargs,
+                endpoint,
+                data,
+                rate_limit_args,
+            ):
+                if TestBaseRateLimiter.rl_received is None:
+                    TestBaseRateLimiter.rl_received = []
+                TestBaseRateLimiter.rl_received.append((endpoint, data, rate_limit_args))
+                return await callback(*args, **kwargs)
+
+        class TestRequest(BaseRequest):
+            async def initialize(self) -> None:
+                pass
+
+            async def shutdown(self) -> None:
+                pass
+
+            async def do_request(self, *args, **kwargs):
+                if TestBaseRateLimiter.request_received is None:
+                    TestBaseRateLimiter.request_received = []
+                TestBaseRateLimiter.request_received.append((args, kwargs))
+                # return bot.bot.to_dict() for the `get_me` call in `Bot.initialize`
+                return 200, json.dumps({"ok": True, "result": bot.bot.to_dict()}).encode()
+
+        defaults = Defaults(parse_mode=ParseMode.HTML)
+        test_request = TestRequest()
+        standard_bot = ExtBot(token=bot.token, defaults=defaults, request=test_request)
+        rl_bot = ExtBot(
+            token=bot.token,
+            defaults=defaults,
+            request=test_request,
+            rate_limiter=TestRateLimiter(),
+        )
+
+        async with standard_bot:
+            await standard_bot.set_my_commands(
+                commands=[BotCommand("test", "test")], language_code="en"
+            )
+        async with rl_bot:
+            await rl_bot.set_my_commands(
+                commands=[BotCommand("test", "test")],
+                language_code="en",
+                rate_limit_args=(43, "test-1"),
```

Comment:
shouldn't we also test passing api_kwargs here and then checking it's presence in `BaseRateLimiter.data`?

Reply:
yes, we probably should :)

Reply:
test for ptbgithubbridge - ignore me